### PR TITLE
Refactor CastVote deserializer

### PR DESCRIPTION
**Pull request by ahominal**
- added cleaner implementation of the custom CastVote deserializer

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.lang.reflect.Type;
+import java.util.List;
+
+public class JsonCastVoteDeserializer implements JsonDeserializer<CastVote> {
+
+    @Override
+    public CastVote deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+        JsonObject obj = json.getAsJsonObject();
+        Log.d("laoIdField is:", "dsddd");
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:24]: Should be removed I imagine

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        long createdAt = context.deserialize(createdAtField, long.class);
+
+        boolean typeValidationInt = true;
+        boolean typeValidationString = true;
+        // Vote type of a CastVote is either an integer for an OpenBallot election or a
+        // String for an Encrypted election, type should be valid for all votes
+        for (int i = 0; i < jsonVote.size(); i++) {
+            JsonObject voteContent = jsonVote.get(i).getAsJsonObject();
+            typeValidationInt =
+                    typeValidationInt && voteContent.get("vote").getAsJsonPrimitive().isNumber();
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:44]: field should use a constant as key

**Review by Warhedgehog — CHANGES_REQUESTED** [review]
Couple things I'd like to be addressed

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            typeValidationString =
+                    typeValidationString && voteContent.get("vote").getAsJsonPrimitive().isString();
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:46]: Same as above

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 4]
```diff
@@ -38,12 +37,13 @@ public CastVote deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
         boolean typeValidationString = true;
         // Vote type of a CastVote is either an integer for an OpenBallot election or a
         // String for an Encrypted election, type should be valid for all votes
+        String voteIndex = "vote";
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:40]: It should be a public static final String of the class

**Review by Warhedgehog — CHANGES_REQUESTED** [review]

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java (hunk 4 above)
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:40]: ohhh I see, why not private static, I think we won't use it anywhere else

Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java]: Then private is fine

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 5]
```diff
@@ -18,6 +17,8 @@
 
 public class JsonCastVoteDeserializer implements JsonDeserializer<CastVote> {
 
+    private static String voteIndex = "vote";
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:20]:
```suggestion
    private static final String VOTE_INDEX = "vote";
```

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        long createdAt = context.deserialize(createdAtField, long.class);
+
+        boolean typeValidationInt = true;
+        boolean typeValidationString = true;
+        // Vote type of a CastVote is either an integer for an OpenBallot election or a
+        // String for an Encrypted election, type should be valid for all votes
+        for (int i = 0; i < jsonVote.size(); i++) {
+            JsonObject voteContent = jsonVote.get(i).getAsJsonObject();
+            typeValidationInt =
+                    typeValidationInt && voteContent.get(voteIndex).getAsJsonPrimitive().isNumber();
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:44]:
```suggestion
                    typeValidationInt && voteContent.get(VOTE_INDEX).getAsJsonPrimitive().isNumber();
```

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            typeValidationString =
+                    typeValidationString && voteContent.get(voteIndex).getAsJsonPrimitive().isString();
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:46]:
```suggestion
                    typeValidationString && voteContent.get(VOTE_INDEX).getAsJsonPrimitive().isString();
```

**Review by Warhedgehog — CHANGES_REQUESTED** [review]

**Review by zoemarin — APPROVED** [review]
The changes looks fine to me

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                    typeValidationString && voteContent.get(VOTE_INDEX).getAsJsonPrimitive().isString();
+        }
+        if (typeValidationInt && !typeValidationString) {
+            Type token = new TypeToken<List<ElectionVote>>() {}.getType();
+            List<ElectionVote> votes = context.deserialize(jsonVote, token);
+            return new CastVote(votes, electionId, laoId, createdAt);
+        } else if (!typeValidationInt && typeValidationString) {
+            Type token = new TypeToken<List<ElectionEncryptedVote>>() {}.getType();
+            List<ElectionEncryptedVote> votes = context.deserialize(jsonVote, token);
+            return new CastVote(votes, electionId, laoId, createdAt);
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonCastVoteDeserializer.java:55]: That's an elegant way to fix the issue.

Great job !

**Review by GabrielFleischer — APPROVED** [review]
LGTM

**Review by Warhedgehog — APPROVED** [review]
LGTM

**Pull request merged**